ESP8266: Avoid duplicate data sends

michalpasztamobica opened the pull request:
<!--
For more information on the requirements for pull requests, please see [the contributing guidelines](https://os.mbed.com/docs/mbed-os/latest/contributing/workflow.html).

NOTE: Do not remove any of the template headings (even for optional sections) as this
template is automatically parsed. 
-->

### Summary of changes <!-- Required -->

Fixes https://github.com/ARMmbed/mbed-os/issues/11544.

EDIT 30.12.2019: Changed the description quite heavily, therefore did not leave the old part ~~stroke through~~, see history if you want to see the original description.

EDIT 9.01.2020: Now I updated the description below.

The original issue report points out that in case the ESP8266 fails to respond within serial timeout period (2 seconds), there is no way of telling if the packet was sent or not. Therefore if the application decides to retry sending it might be that packet gets sent out twice. 

Following the remarks from @ccli8 and from @kjbracey-arm I propose the following:

1) ESP8266 driver checks if the undocumented `Recvd x bytes` message arrives. This basically acknowledges successful UART transfer. If it doesn't come or if any other serial-related error pops up, we can assume the ESP is an unknown state and better be reset - we return `DEVICE_ERROR` if only the `SEND OK` is missing, we return `WOULD_BLOCK` and can handle the acknowledgement asynchronously via an OOB.

~~2) If the serial acknowledge arrives, we can wait a while for `SEND OK` message (although, according to @kjbracey-arm 's explanation, we don't really have to). There must be some timeout to this wait. I added a counter that allows 3 blocking checks for `SEND OK` in case of `busy s...` message.
@AnttiKauppila , I moved your recently added retry mechanism waiting for `SEND OK`. Now that we check for serial acknowledgement it makes sense to do this wait a bit later. I also added a limit to it (3 retries), please bear this in mind during review.~~

2) @ccli8 has proposed more fixes, so this PR also fixes the reopening of the socket (with the possible `SEND FAIL` in mind, introduces per-socket failure flag and removed the large chunk of code which was holding `send()` back waiting for `busy` or `SEND OK/FAIL` - @AnttiKauppila , please pay attention to this part when reviewing as this removes your recent changes.

I checked that all greentea tests which are passing on master are also passing with this PR (CI is not running ESP8266) and the `mbed-os-example-pelion` connects fine. I noticed however that netsocket-udp test suite takes much longer to pass in RAAS (more than 1100 seconds), so the timeout was increased accordingly.

<!-- 
    Please provide the following information: 

    Description of the the change (what is this fixing / adding / removing?).

    Why the change is needed (if this is fixing a reported issue please summarize what
    the issue is and add the reference. E.g. Fixes #17119).

    NOTE: This section is automatically written to release notes for Feature and 
    Major releases and should contain enough details for a user.
    
-->

#### Impact of changes <!-- Optional -->

The `send()` calls might take longer, but success rate will increase.

#### Migration actions required <!-- Optional -->

None

### Documentation <!-- Required -->

None

----------------------------------------------------------------------------------------------------------------
### Pull request type <!-- Required -->

<!--
    Please add only one X to one of the following types. Do not fill multiple types (split the pull request otherwise).
    Please note this is not a GitHub task list, indenting the boxes or changing the format to add a '.' or '*' in front of them would change the meaning incorrectly. 
-->
    [x] Patch update (Bug fix / Target update / Docs update / Test update / Refactor)
    [] Feature update (New feature / Functionality change / New API)
    [] Major update (Breaking change E.g. Return code change / API behaviour change)

----------------------------------------------------------------------------------------------------------------
### Test results <!-- Required -->

<!--
    Provide all the information required, listing all the testing performed. For new targets please attach full test results for all supported compilers.
-->
    [] No Tests required for this change (E.g docs only update)
    [x] Covered by existing mbed-os tests (Greentea or Unittest)
    [] Tests / results supplied as part of this PR
    
    
----------------------------------------------------------------------------------------------------------------
### Reviewers <!-- Optional -->

@ccli8 
@AnttiKauppila 
@VeijoPesonen 
@SeppoTakalo 

----------------------------------------------------------------------------------------------------------------

michalpasztamobica committed:
ESP8266: add a retry mechanism to avoid duplicate data sends

We are now checking if ESP8266 has confirmed receiving data over serial
port with an undocumented (but existing) "Recv x bytes" message. Next we
are explicitly waiting for an official "SEND OK".

ccli8 commented on components/wifi/esp8266-driver/ESP8266/ESP8266.cpp:
```diff
@@ -655,7 +658,34 @@ nsapi_error_t ESP8266::send(int id, const void *data, uint32_t amount)
-    if (_parser.write((char *)data, (int)amount) >= 0 && _parser.recv("SEND OK")) {
+    if (_parser.write((char *)data, (int)amount) < 0) {
+        tr_debug("Failed to write data");
+        ret = NSAPI_ERROR_DEVICE_ERROR;
```
Missing `goto END`?

ccli8 commented on components/wifi/esp8266-driver/ESP8266/ESP8266.cpp:
```diff
@@ -655,7 +658,34 @@ nsapi_error_t ESP8266::send(int id, const void *data, uint32_t amount)
+    int bytes_confirmed;
+    if (!_parser.recv("Recv %d bytes", &bytes_confirmed)) {
+        tr_debug("Bytes not confirmed.");
+        ret = NSAPI_ERROR_DEVICE_ERROR;
```
Missing `goto END`?

ccli8 commented on components/wifi/esp8266-driver/ESP8266/ESP8266.cpp:
```diff
@@ -655,7 +658,34 @@ nsapi_error_t ESP8266::send(int id, const void *data, uint32_t amount)
+        ret = NSAPI_ERROR_DEVICE_ERROR;
+    } else if (bytes_confirmed != amount) {
+        tr_debug("Error: confirmed %d bytes, but expected %d.", bytes_confirmed, amount);
+        ret = NSAPI_ERROR_DEVICE_ERROR;
```
Missing `goto END`?

michalpasztamobica committed:
ESP8266: Improve error handling and partial sends

VeijoPesonen commented on components/wifi/esp8266-driver/ESP8266/ESP8266.cpp:
```diff
@@ -643,20 +644,63 @@ nsapi_error_t ESP8266::send(int id, const void *data, uint32_t amount)
+        goto END;
+    }
+
+    send_ack_retries = 3;
```
Unnecessary because the variable has already been initialized earlier.

VeijoPesonen commented on components/wifi/esp8266-driver/ESP8266/ESP8266.cpp:
```diff
@@ -678,9 +722,9 @@ nsapi_error_t ESP8266::send(int id, const void *data, uint32_t amount)
     }
 
-    if (!_sock_i[id].open && ret != NSAPI_ERROR_OK) {
+    if (!_sock_i[id].open && ret < NSAPI_ERROR_OK) {
```
Smaller than zero would be better because we know that negative values are error codes.

VeijoPesonen commented on components/wifi/esp8266-driver/ESP8266/ESP8266.cpp:
```diff
@@ -643,20 +644,63 @@ nsapi_error_t ESP8266::send(int id, const void *data, uint32_t amount)
-    if (_parser.write((char *)data, (int)amount) >= 0 && _parser.recv("SEND OK")) {
-        ret = NSAPI_ERROR_OK;
+    if (_parser.write((char *)data, (int)amount) < 0) {
+        tr_debug("Failed to write serial data");
```
Convention has been to prefix a trace with the name of the function from which it gets emitted.

VeijoPesonen commented on components/wifi/esp8266-driver/ESP8266/ESP8266.cpp:
```diff
@@ -643,20 +644,63 @@ nsapi_error_t ESP8266::send(int id, const void *data, uint32_t amount)
+        if (_busy) {
+            _busy = false;
+            tr_debug("Busy, retry...");
+            goto RETRY_SEND_ACK;
```
Maybe you should use a for-loop instead?

michalpasztamobica replied:
True... I tried to stick to the `goto`-style too much.

ccli8 commented on components/wifi/esp8266-driver/ESP8266/ESP8266.cpp:
```diff
@@ -612,9 +614,18 @@ bool ESP8266::dns_lookup(const char *name, char *ip)
+    if (_prev_send_ok_pending && _sock_i[id].proto == NSAPI_TCP) {
+        tr_debug("send(): Previous packet was not ACK-ed with SEND OK.");
+        return NSAPI_ERROR_WOULD_BLOCK;
+    }
```
Is it necessary to trigger `ATCmdParser` to roll, like calling `process_oob`? Otherwise, OOB message `SEND OK` is not handled.

michalpasztamobica replied:
@ccli8 , please correct me if I am wrong, but it seems that sigio will automatically trigger oob processing.

`ESP8266`'s sigio is [connected to serial's sigio](https://github.com/ARMmbed/mbed-os/blob/5f495db06bd6c6e0ac0adf001be57f1d724f9aa1/components/wifi/esp8266-driver/ESP8266/ESP8266.cpp#L1001).
`ESP8266Interface`'s sigio is [connected to that of](https://github.com/ARMmbed/mbed-os/blob/5f495db06bd6c6e0ac0adf001be57f1d724f9aa1/components/wifi/esp8266-driver/ESP8266Interface.cpp#L132) `ESP8266`.
So if anything comes over serial it will trigger `ESP8266Interface::event`, [which in turn will schedule](https://github.com/ARMmbed/mbed-os/blob/5f495db06bd6c6e0ac0adf001be57f1d724f9aa1/components/wifi/esp8266-driver/ESP8266Interface.cpp#L1028) `ESP8266Interface::proc_oob_evnt()` (unless there is one scheduled already).
That [will trigger oob processing](https://github.com/ARMmbed/mbed-os/blob/5f495db06bd6c6e0ac0adf001be57f1d724f9aa1/components/wifi/esp8266-driver/ESP8266Interface.cpp#L1124).

Does this make sense?

ccli8 replied:
@michalpasztamobica Thanks for your explanation. That makes sense.

Still have other concern with it. When some exception like disconnect or error happens, `_prev_send_ok_pending` doesn't get cleared and program will block here indefinitely. With `_prev_send_ok_pending` and `_send_fail_received` mutual exclusion, plus other exception cases with send ack, recommend combining `_prev_send_ok_pending` and `_send_fail_received` into one which has states like: no send, send pending, send ok acked, and send fail acked. One point is that program cannot block here in send ack exception cases.

```C++
    if (_prev_send_ok_pending && _sock_i[id].proto == NSAPI_TCP) {
        tr_debug("send(): Previous packet was not ACK-ed with SEND OK.");
        return NSAPI_ERROR_WOULD_BLOCK;
    }
```

Potential memory leak with new oobs:
```C++
_parser.oob("SEND FAIL", callback(this, &ESP8266::_oob_send_fail_received));
_parser.oob("SEND OK", callback(this, &ESP8266::_oob_send_ok_received));
```
`_parser.oob` is not easy to pair `_parser.remove_oob` in exception cases. Recommend just placing them, like other oobs, in `ESP8266::ESP8266` constructor and modifying related logic.

VeijoPesonen commented on components/wifi/esp8266-driver/ESP8266/ESP8266.cpp:
```diff
@@ -635,52 +645,91 @@ nsapi_error_t ESP8266::send(int id, const void *data, uint32_t amount)
+    // The "Recv X bytes" is not documented.
+    if (!_parser.recv("Recv %d bytes", &bytes_confirmed)) {
+        tr_debug("send(): Bytes not confirmed.");
+        ret = NSAPI_ERROR_DEVICE_ERROR;
```
This line should be removed because TCP and UDP are the only protocols supported.

VeijoPesonen commented on components/wifi/esp8266-driver/ESP8266/ESP8266.cpp:
```diff
@@ -635,52 +645,91 @@ nsapi_error_t ESP8266::send(int id, const void *data, uint32_t amount)
+
+    // ESP8266 ACKed data over serial, but did not ACK over TCP or report any error.
+    _prev_send_ok_pending = true;
+    _parser.oob("SEND OK", callback(this, &ESP8266::_oob_send_ok_received));
```
I see OOB handler added here but I don't see one removed anywhere in the code.

michalpasztamobica committed:
ESP8266: Introduce single send_status instead of two flags

michalpasztamobica committed:
Increase UDP test timeout to 1200s

This is because the ESP8266 is now waiting for SEND OK and takes much
more to complete the send_repeat, and echo_burst tests in RAAS.
(but has higher pass ratio).

ccli8 committed:
ESP8266: Simplify flow control and enable per-socket reconnection

1.  Fix 'spurious close' by adding close() in open(). 'spurious close' gets frequent and cannot ignore when send() changes to asynchronous. User can retry open() until 'spurious close' gets true.
2.  Allow only one actively sending socket because:
    (1) ESP8266 AT packets 'SEND OK'/'SEND FAIL' are not associated with socket ID. No way to tell them.
    (2) In original implementation, ESP8266::send() is synchronous, which implies only one actively sending socket.
3.  Register 'SEND OK'/'SEND FAIL' oobs, like others in ESP8266::ESP8266 constructor. Don't get involved in oob management with send status because ESP8266 modem possibly doesn't reply these packets on error case.
4.  Now that ESP8266::send() changes to asynchronous, drop the code with _parser.recv("SEND OK")/_parser.recv("SEND FAIL"). _parser.recv("SEND OK")/_parser.recv("SEND FAIL") and 'SEND OK'/'SEND FAIL' oobs both consume 'SEND OK'/'SEND FAIL' packets and complicate flow control.

VeijoPesonen commented on components/wifi/esp8266-driver/ESP8266/ESP8266.cpp:
```diff
@@ -964,20 +1020,33 @@ bool ESP8266::close(int id)
+                if (id == _sock_sending_id) {
+                    _sock_sending_id = -1;
+                }
+                _sock_i[id].send_fail = false;
```
This same snippet has been added at least to 6 different places so maybe it should be turned into a function.

michalpasztamobica replied:
Added a function and replaced the reoccurent code.

VeijoPesonen commented on components/wifi/esp8266-driver/ESP8266/ESP8266.cpp:
```diff
@@ -612,9 +630,24 @@ bool ESP8266::dns_lookup(const char *name, char *ip)
+        if (_sock_sending_id >= 0 && _sock_sending_id < SOCKET_COUNT) {
+            if (!_sock_i[id].send_fail) {
+                tr_debug("send(): Previous packet (socket %d) was not yet ACK-ed with SEND OK.", _sock_sending_id);
+                return NSAPI_ERROR_WOULD_BLOCK;
```
What if modem doesn't reply with anything? Might `SEND OK` or `SEND FAIL` get lost? I'm just thinking out loud here so this isn't a request to change anything.

michalpasztamobica replied:
This was discussed in the main thread. Please [see this comment](https://github.com/ARMmbed/mbed-os/pull/12157#issuecomment-570234338) and the previous one.
Long story short: `ERROR` is another possibility aside from `SEND OK` and `SEND FAIL`, but as experiments showed it is a recoverable error, so we basically ignore it.
If neither `SEND OK` nor `SEND FAIL` are coming we just keep returning `WOULD_BLOCK` to any new `send()` attempts. It's up to the application to decide how long this can be tolerated. I assume some socket timeout will take care of this in a typical mbed application?

VeijoPesonen commented on components/wifi/esp8266-driver/ESP8266/ESP8266.cpp:
```diff
@@ -635,52 +671,72 @@ nsapi_error_t ESP8266::send(int id, const void *data, uint32_t amount)
+        goto END;
+    } else if (bytes_confirmed != amount) {
+        tr_debug("send(): Error: confirmed %d bytes, but expected %d.", bytes_confirmed, amount);
+        ret = NSAPI_ERROR_DEVICE_ERROR;
```
Would it be ok in case of TCP that ESP8266 accepts less data than we are trying to send?

michalpasztamobica replied:
I thought about this as well, but wasn't brave enough to implement it ;-). I have never seen this situation to happen, so I assume it's either all or nothing, but you are right, if ESP ever decides to accept partial write, we should be able to handle it.

michalpasztamobica committed:
ESP8266: accept partial writes for TCP and clean up code
